File: Core/src/main/java/jokatu/game/event/AbstractEventHandler.java
package jokatu.game.event;

import jokatu.game.Game;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;

/**
 * Listens to events but ignores the ones that aren't for the game and the event type it specifies.
 * @param <G> the type of {@link Game} to accept
 * @param <E> the type of {@link GameEvent} to accept
 */
@Component
public abstract class AbstractEventHandler<G extends Game, E extends GameEvent> extends SpecificEventHandler<E> {

	@Override
	protected final void handleCastEvent(@NotNull Game<?> game, @NotNull E event) {
		if (getGameClass().isInstance(game)) {
			handleCastGameAndEvent(getGameClass().cast(game), event);
		}
	}

	@NotNull
	protected abstract Class<G> getGameClass();

	protected abstract void handleCastGameAndEvent(@NotNull G game, @NotNull E event);
}

File: Core/src/main/kotlin/jokatu/game/event/AbstractEventHandler.kt
package jokatu.game.event

import jokatu.game.Game
import kotlin.reflect.KClass
import kotlin.reflect.full.cast

/**
 * Listens to events but ignores the ones that aren't for the game and the event type it specifies.
 * @param <G> the type of [Game] to accept
 * @param <E> the type of [GameEvent] to accept
 */
abstract class AbstractEventHandler<G : Game<*>, E : GameEvent>(
		private val gameClass: KClass<G>,
		eventClass: KClass<E>
) : SpecificEventHandler<E>(eventClass) {
	override fun handleCastEvent(game: Game<*>, event: E) {
		if (gameClass.isInstance(game)) {
			handleCastGameAndEvent(gameClass.cast(game), event)
		}
	}

	protected abstract fun handleCastGameAndEvent(game: G, event: E)
}
